File: examples/python/my-app/src/services/user_service.py
"""User service with business logic."""

from typing import List, Optional
from models.user import User
from database.user_repository import UserRepository
from utils.validator import Validator
from utils.logger import Logger
from auth.password_hasher import PasswordHasher

class UserService:
    """Service for user-related business logic."""
    
    def __init__(self):
        self._repository = UserRepository()
        self._validator = Validator()
        self._logger = Logger.get_instance()
        self._password_hasher = PasswordHasher()
    
    def get_user_by_id(self, user_id: int) -> Optional[User]:
        """Get user by ID."""
        self._logger.info(f"Getting user: {user_id}")
        return self._repository.find_by_id(user_id)
    
    def get_all_users(self) -> List[User]:
        """Get all users."""
        self._logger.info("Getting all users")
        return self._repository.find_all()
    
    def create_user(self, username: str, email: str, password: str, full_name: str) -> User:
        """Create a new user."""
        self._logger.info(f"Creating user: {username}")
        
        # Validate input
        if not self._validator.is_valid_email(email):
            raise ValueError("Invalid email address")
        
        if not self._validator.is_valid_password(password):
            raise ValueError("Password does not meet requirements")
        
        # Hash password
        password_hash = self._password_hasher.hash_password(password)
        
        # Create user (mock)
        from datetime import datetime
        user = User(
            id=0,
            username=username,
            email=email,
            password_hash=password_hash,
            full_name=full_name,
            created_at=datetime.now()
        )
        
        return self._repository.save(user)
    
    def update_user(self, user: User) -> User:
        """Update existing user."""
        self._logger.info(f"Updating user: {user.id}")
        return self._repository.update(user)
    
    def delete_user(self, user_id: int) -> bool:
        """Delete a user."""
        self._logger.warning(f"Deleting user: {user_id}")
        return self._repository.delete(user_id)

File: examples/delphi/my-app/src/Services/UserService.pas
unit Services.UserService;

interface

uses
  System.SysUtils,
  Models.User,
  Database.UserRepository,
  Utils.Logger,
  Utils.Validator,
  Utils.StringUtils;

type
  TUserService = class
  private
    FRepository: TUserRepository;
    FLogger: TLogger;
    FValidator: TValidator;
    FStringUtils: TStringUtils;
  public
    constructor Create;
    destructor Destroy; override;
    
    function CreateUser(const AUsername, AEmail, APassword: string): TUser;
    function GetUserById(AUserId: Integer): TUser;
    function GetUserByUsername(const AUsername: string): TUser;
    function GetUserByEmail(const AEmail: string): TUser;
    function GetAllUsers: TArray<TUser>;
    procedure UpdateUser(AUser: TUser);
    procedure DeleteUser(AUserId: Integer);
    procedure DeactivateUser(AUserId: Integer);
    procedure ActivateUser(AUserId: Integer);
    function ValidateCredentials(const AUsername, APassword: string): Boolean;
  end;

implementation

constructor TUserService.Create;
begin
  inherited;
  FRepository := TUserRepository.Create;
  FLogger := TLogger.Create;
  FValidator := TValidator.Create;
  FStringUtils := TStringUtils.Create;
end;

destructor TUserService.Destroy;
begin
  FRepository.Free;
  FLogger.Free;
  FValidator.Free;
  FStringUtils.Free;
  inherited;
end;

function TUserService.CreateUser(const AUsername, AEmail, APassword: string): TUser;
var
  PasswordHash: string;
begin
  FLogger.Log('Creating user: ' + AUsername);
  
  if not FValidator.ValidateEmail(AEmail) then
    raise Exception.Create('Invalid email format');
    
  if not FValidator.ValidateString(AUsername, 3, 50) then
    raise Exception.Create('Invalid username');
    
  PasswordHash := FStringUtils.HashPassword(APassword);
  
  Result := TUser.Create;
  Result.SetUsername(AUsername);
  Result.SetEmail(AEmail);
  Result.SetPasswordHash(PasswordHash);
  
  FRepository.Save(Result);
  FLogger.Log('User created with ID: ' + IntToStr(Result.GetId));
end;

function TUserService.GetUserById(AUserId: Integer): TUser;
begin
  Result := FRepository.FindById(AUserId);
end;

function TUserService.GetUserByUsername(const AUsername: string): TUser;
begin
  Result := FRepository.FindByUsername(AUsername);
end;

function TUserService.GetUserByEmail(const AEmail: string): TUser;
begin
  Result := FRepository.FindByEmail(AEmail);
end;

function TUserService.GetAllUsers: TArray<TUser>;
begin
  Result := FRepository.FindAll;
end;

procedure TUserService.UpdateUser(AUser: TUser);
begin
  FLogger.Log('Updating user: ' + IntToStr(AUser.GetId));
  FRepository.Update(AUser);
end;

procedure TUserService.DeleteUser(AUserId: Integer);
begin
  FLogger.Log('Deleting user: ' + IntToStr(AUserId));
  FRepository.Delete(AUserId);
end;

procedure TUserService.DeactivateUser(AUserId: Integer);
var
  User: TUser;
begin
  User := GetUserById(AUserId);
  if Assigned(User) then
  begin
    User.SetIsActive(False);
    UpdateUser(User);
    FLogger.Log('User deactivated: ' + IntToStr(AUserId));
  end;
end;

procedure TUserService.ActivateUser(AUserId: Integer);
var
  User: TUser;
begin
  User := GetUserById(AUserId);
  if Assigned(User) then
  begin
    User.SetIsActive(True);
    UpdateUser(User);
    FLogger.Log('User activated: ' + IntToStr(AUserId));
  end;
end;

function TUserService.ValidateCredentials(const AUsername, APassword: string): Boolean;
var
  User: TUser;
  PasswordHash: string;
begin
  User := GetUserByUsername(AUsername);
  Result := False;
  
  if Assigned(User) and User.GetIsActive then
  begin
    PasswordHash := FStringUtils.HashPassword(APassword);
    Result := User.GetPasswordHash = PasswordHash;
  end;
end;

end.
